powersoftau/src/utils.rs

```diff
@@ -28,6 +29,28 @@ pub fn print_hash(hash: &[u8]) {
     }
 }
 
+/// Exponentiate a large number of points, with an optional coefficient to be applied to the
+/// exponent. Panics if provided arguments have different length.
+pub fn batch_exp<C: CurveAffine>(bases: &mut [C], exps: &[C::Scalar], coeff: Option<&C::Scalar>) {
+    // ensure the 2 vectors have the same length
+    assert_eq!(bases.len(), exps.len());
+    // raise the base to the exponent and assign it back to the base
+    // this will return the points as projective
+    bases.par_iter_mut().enumerate().for_each(|(i, base)| {
+        let mut exp = exps[i];
+
+        // If a coefficient was provided, multiply the exponent
+        // by that coefficient
+        if let Some(coeff) = coeff {
+            exp.mul_assign(coeff);
+        }
+
+        // Raise the base to the exponent (additive notation so it is executed
+        // via a multiplication)
+        *base = base.mul(exp).into_affine()
```

Comment:
We should instead do batch normalization on everything afterwards.

Reply:
fixed in https://github.com/celo-org/snark-setup/pull/5/commits/0e017ff501e7d8af44c13bd74cb49c2f16679091